### PR TITLE
sct(component): add `TimestampByHeight` to query service

### PR DESCRIPTION
**Pull request by turbocrime**
## Describe your changes

Adds a TimestampByHeightRequest to SCT QueryService.

There is no corresponding migration added, so any block heights lower than the height at which this is deployed will have missing data.

## Issue ticket number and link

https://github.com/penumbra-zone/penumbra/issues/4522

## Checklist before requesting a review

- [ ] If this code contains consensus-breaking changes, I have added the "consensus-breaking" label. Otherwise, I declare my belief that there are not consensus-breaking changes, for the following reason:

  > This adds a pb API endpoint and should not affect consensus. New data is tracked in nonconsensus storage only.

### CODE REVIEW

**Comment by turbocrime** [comment]
no tests - haven't looked at that yet. is this the correct format of the timestamp? are there other considerations to make?

File: crates/core/component/sct/src/component/rpc.rs [hunk 1]
```diff
@@ -55,4 +56,24 @@ impl QueryService for Server {
...
+
+    #[instrument(skip(self, request))]
+    async fn timestamp_by_height(
+        &self,
+        request: tonic::Request<TimestampByHeightRequest>,
+    ) -> Result<tonic::Response<TimestampByHeightResponse>, Status> {
+        let state = self.storage.latest_snapshot();
+
+        let height = request.get_ref().height;
+        let block_time = state.get_block_timestamp().await.map_err(|e| {
```
Comment (erwanor) [inline crates/core/component/sct/src/component/rpc.rs:68]: This will always return the timestamp for the latest block in the snapshot

Comment (erwanor) [reply crates/core/component/sct/src/component/rpc.rs:68]: I don't think we store height-by-height timestamp, but I can take a look once I'm back at the computer

Comment (zbuc) [reply crates/core/component/sct/src/component/rpc.rs:68]: Correct, we would need to add a new nonverifiable storage key containing timestamp-by-height

**Comment by zbuc** [comment]
I added a modification to track historic timestamps. Unfortunately writing a migration for this isn't really straightforward, but it's probably fine that historic data won't exist as it's in non-verifiable storage.

**Comment by aubrika** [comment]
Seems like as written, worst case scenario is that querying for a historical block timestamp without a migration to populate that data is just going to throw an error. @turbocrime is that acceptable for your intended use case - i.e. erroring when called on blocks prior to an upgrade, returning data afterwards?

**Comment by turbocrime** [comment]
throwing is fine - it's a non-critical case and we have a fallback already

**Review by turbocrime — COMMENTED** [review]
reviewed, looks good

**Review by zbuc — APPROVED** [review]

**Pull request merged**